File: C++/valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        unordered_map<char, char> charMap{
            {')', '('},
            {']', '['},
            {'}', '{'}
        };

        for (char c : s){
            if (charMap.contains(c) && (st.empty() || st.top() != charMap[c])){return false;}
            else if(charMap.contains(c)){st.pop();}
            else{st.push(c);}
        }

        return st.empty();
    }
};

File: python/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        charMap = {
            ')':'(',
            ']':'[',
            '}':'{'
        }

        for c in s:
            if c in charMap and (not stack or stack[-1] != charMap[c]):
                return False
            elif c in charMap:
                stack.pop()
            else:
                stack.append(c)
        return not stack
            
